prqlc/prql-compiler/src/sql/dialect.rs

```diff
@@ -185,12 +212,17 @@ pub(super) trait DialectHandler: Any + Debug {
     /// (see https://docs.rs/chrono/latest/chrono/format/strftime/index.html)
     fn translate_prql_date_format(&self, prql_date_format: &str) -> Result<String> {
         Ok(StrftimeItems::new(prql_date_format)
-            .map(|item| self.translate_chrono_item(item))
+            .map(|item| {
+                if !is_supported_chrono_item(&item) {
+                    bail!("PRQL doesn't support this date format item")
+                }
+                self.translate_chrono_item(&item)
+            })
```

Comment:
I think the more rust-y thing to do here is to change the `translate_chrono_item` `_` branch to a `bail!`, remove the `is_supported_chrono_item` check, and then it should all flow through.

Then we also don't need to have two similar lists. Does that make sense?

Reply:
For me it's not a question of rusty or not rusty but more of "where is it decided that it's supported"
And in my mind it was in the `is_supported_chrono_item` and we could write the documentation based on those specifiers.
Even if a dialect can do more, we don't support it natively in PRQL (people can always do a `s-string`)
This was a way to have exactly the same body for all the `translate_chrono_item` methods for all the dialects

I can change it if you want but this is why I went with this approach

Reply:
If we don't do that, we need to `bail!` with the same error message for all the dialects. And make sure they implement exactly the same specifiers.
But maybe I'm taking the wrong approach or there is a nicer way.
Let me know what you think

Reply:
The problem with the currently proposed implementation is that if we make a small mistake — i.e. the code in `is_supported_chrono_item` covers a slightly different set of cases than a dialect's `translate_chrono_item` — then we get a panic (and we should never panic, that's always a bug). So we lose exhaustiveness — a big advantage of pattern matching.

> If we don't do that, we need to `bail!` with the same error message for all the dialects.

The approach I'm thinking is that we can write any message we want in each dialect (e.g. at https://github.com/PRQL/prql/pull/3954/files#diff-65a3c21d281e36b6fff6cc3e5ffdb821f86f3a54388bf07f67fdde4c17f6e90dR378). And we don't have any `is_supported_chrono_item` function.

> And make sure they implement exactly the same specifiers.

I agree, this requires some coordination. But IIUC — we have to do that now; it's just that if we make a mistake, we get a panic. (Very possible I'm misunderstanding, let me know)

> And in my mind it was in the `is_supported_chrono_item` and we could write the documentation based on those specifiers.

I agree we lose the canonical set of identifies — that's a disadvantage (although it wasn't enforced previously, it was just somewhere to write them). I think probably the canonical source can be a query which is then tested across every dialect + serves as documentation (we don't currently run our docs through every dialect, but we could quite easily add a [tag] which would then run the query over every dialect, or just copy & paste with our integration tests for the moment...)

WDYT?

Reply:
Thank you very much for your feedback. All the specifiers that I added in the doc are tested in the integration tests so ok let's do that. I'll go back to the approach with one bail per dialect :+1: